### PR TITLE
lmstudio: 0.2.22 -> 0.2.23

### PR DESCRIPTION
**Pull request by cig0**
## Description of changes

- Upstream minor patch update update (0.2.22.x)
- Updated download links

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](.))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by cig0** [comment]
Hi everyone,
This is my first attempt to contribute to `Nixpkgs`.
While I thoroughly read the contribution guidelines, this PR will undoubtedly be lacking in many ways - I beg you all for patience.
I'm looking forward to learning from your observations!
Thanks.

**Comment by drupol** [comment]
Hi!

Thanks for your first contribution. Technically, it looks good.
However, is it really necessary to update to 0.2.22c? Can't we just wait for the next version?

**Comment by cig0** [comment]
> Hi!
> 
> Thanks for your first contribution. Technically, it looks good. However, is it really necessary to update to 0.2.22c? Can't we just wait for the next version?

Hi @drupol thanks for the quick reply -- and for maintaining this package!

I believe I failed to provide a better - more concise - context on why I opened this PR:
- The current package is failing to install LM Studio (please see full log below)
- I also noticed LM Studio devs also changed the download URL

```
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.8.9, NixOS, 24.05 (Uakari), 24.05pre623656.f1010e0469db`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.2`
 - channels(root): `"nixos, nixos-hardware"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

$ nix-shell -p lmstudio
these 5 derivations will be built:
  /nix/store/n394c0bgqizy2xgjxlmi7nainj7mvqqd-LM_Studio-0.2.22.AppImage.drv
  /nix/store/jv7k3czv5c7y47cvb2irj6wjhy81l240-lmstudio-0.2.22-extracted.drv
  /nix/store/dqdly8xpyikns094p4xic7lhhpblagh9-lmstudio-0.2.22-init.drv
  /nix/store/91dhzjwgh62k82fr72vpk60yjr2iygvr-lmstudio-0.2.22-bwrap.drv
  /nix/store/s35za31rmv18xswbmhxbraw5bbwrww1n-lmstudio-0.2.22.drv
building '/nix/store/n394c0bgqizy2xgjxlmi7nainj7mvqqd-LM_Studio-0.2.22.AppImage.drv'...

trying https://releases.lmstudio.ai/linux/0.2.22/beta/LM_Studio-0.2.22.AppImage
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  545M  100  545M    0     0  3968k      0  0:02:20  0:02:20 --:--:-- 4356k
error: hash mismatch in fixed-output derivation '/nix/store/n394c0bgqizy2xgjxlmi7nainj7mvqqd-LM_Studio-0.2.22.AppImage.drv':
         specified: sha256-hcV8wDhulFAxHDBDKicpEGovwcsn9RaIi/idUz+YzD8=
            got:    sha256-HoAVUU5Z5ueGMSybhT0OKV611v1yU5am0wy0GsCUmvk=
error: 1 dependencies of derivation '/nix/store/jv7k3czv5c7y47cvb2irj6wjhy81l240-lmstudio-0.2.22-extracted.drv' failed to build
error: 1 dependencies of derivation '/nix/store/s35za31rmv18xswbmhxbraw5bbwrww1n-lmstudio-0.2.22.drv' failed to build
```

**Comment by pluiedev** [comment]
> 
> $ nix-shell -p lmstudio
> 
> these 5 derivations will be built:
> 
>   /nix/store/n394c0bgqizy2xgjxlmi7nainj7mvqqd-LM_Studio-0.2.22.AppImage.drv
> 
>   /nix/store/jv7k3czv5c7y47cvb2irj6wjhy81l240-lmstudio-0.2.22-extracted.drv
> 
>   /nix/store/dqdly8xpyikns094p4xic7lhhpblagh9-lmstudio-0.2.22-init.drv
> 
>   /nix/store/91dhzjwgh62k82fr72vpk60yjr2iygvr-lmstudio-0.2.22-bwrap.drv
> 
>   /nix/store/s35za31rmv18xswbmhxbraw5bbwrww1n-lmstudio-0.2.22.drv
> 
> building '/nix/store/n394c0bgqizy2xgjxlmi7nainj7mvqqd-LM_Studio-0.2.22.AppImage.drv'...
> 
> 
> 
> trying https://releases.lmstudio.ai/linux/0.2.22/beta/LM_Studio-0.2.22.AppImage
> 
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
> 
>                                  Dload  Upload   Total   Spent    Left  Speed
> 
> 100  545M  100  545M    0     0  3968k      0  0:02:20  0:02:20 --:--:-- 4356k
> 
> error: hash mismatch in fixed-output derivation '/nix/store/n394c0bgqizy2xgjxlmi7nainj7mvqqd-LM_Studio-0.2.22.AppImage.drv':
> 
>          specified: sha256-hcV8wDhulFAxHDBDKicpEGovwcsn9RaIi/idUz+YzD8=
> 
>             got:    sha256-HoAVUU5Z5ueGMSybhT0OKV611v1yU5am0wy0GsCUmvk=
> 
> error: 1 dependencies of derivation '/nix/store/jv7k3czv5c7y47cvb2irj6wjhy81l240-lmstudio-0.2.22-extracted.drv' failed to build
> 
> error: 1 dependencies of derivation '/nix/store/s35za31rmv18xswbmhxbraw5bbwrww1n-lmstudio-0.2.22.drv' failed to build
> 
> ```

A hash mismatch? Now that is *very* odd. That means upstream is now delivering a different binary compared to before which is generally a pretty bad sign — if they can just change it willy-nilly, it potentially opens up downstream users for supply chain attacks. Bit hyperbolical, but it could happen.

**Comment by Luk45135** [comment]
> A hash mismatch? Now that is _very_ odd. That means upstream is now delivering a different binary compared to before which is generally a pretty bad sign — if they can just change it willy-nilly, it potentially opens up downstream users for supply chain attacks. Bit hyperbolical, but it could happen.

But isn't that always the case if it's not hosted on github. They can just change what binary is behind that link (and even if it is open source the maintainers would have to read every commit to verify that there is no malware which just isn't feasible.) Isn't that the reason we have this system to verify the hash before installing in the first place? Even then it's possible to change the binary without us knowing with a hash collision attack. To completely alleviate this issue we would have to host every binary on our own, which also isn't feasible without major server costs. I think this way of doing things is just the tradeoff we have landed on.

**Comment by pluiedev** [comment]
> But isn't that always the case if it's not hosted on github. They can just change what binary is behind that link (and even if it is open source the maintainers would have to read every commit to verify that there is no malware which just isn't feasible.) Isn't that the reason we have this system to verify the hash before installing in the first place? Even then it's possible to change the binary without us knowing with a hash collision attack. To completely alleviate this issue we would have to host every binary on our own, which also isn't feasible without major server costs. I think this way of doing things is just the tradeoff we have landed on. 
> 
> But i do think that it's odd that they changed the binary that's behind the link.

Yeah, that is indeed the point, and we're not even allowed to host these binaries as they are unfree and not redistributable. However, it is quite worrying that they would just swap out the binary like that, which makes this source not reliable enough for our purposes. Stuff packaged in Nixpkgs has to build reliably and correctly, and that having a reliable source is a key part of that.

**Comment by pluiedev** [comment]
Looks like [upstream has pledged to keep the download links stable](https://github.com/lmstudio-ai/lmstudio-bug-tracker/issues/25#issuecomment-2123475939) for now. Let's just fix the link and the hash and get on with it now, then?

**Comment by eeedean** [comment]
@cig0 will you update your remote branch to point to the most recent version?

**Comment by cig0** [comment]
@eeedean @drupol Heads up: LM Studio devs updated the GNU/Linux filename package, adding `-Ubuntu-20.04`: https://github.com/NixOS/nixpkgs/pull/310914/commits/39adf42f96fdea869664ae532eb32f19d6a3e4a7

At first, I thought about adding a variable to handle the new string, but I don't feel it'll change for a while since 20.04 is an LTS release.

I thought about letting you know before proceeding with the change —you guys are the package maintainers- but I preferred agility over bureaucracy here, as this is a minimal change.

Let me know what you think and if you prefer moving the OS bit to its own variable.

Thanks!

**Comment by drupol** [comment]
I have no strong opinion on this. I don't really use LMStudio anymore.

Can I ask you if you could remove my name from the maintainers list?

File: pkgs/by-name/lm/lmstudio/darwin.nix [hunk 1]
```diff
@@ -10,8 +10,8 @@ stdenv.mkDerivation {
   inherit meta pname version;
 
   src = fetchurl {
-    url = "https://releases.lmstudio.ai/mac/arm64/${version}/b/latest/LM-Studio-${version}-arm64.dmg";
-    hash = "sha256-kb1XoTZjhCL1+CsV+r3/EN0srzIJ43H2VMZ779dVq1k=";
+    url = "https://releases.lmstudio.ai/mac/arm64/${version}/e/latest/LM-Studio-${version}-arm64.dmg";
```
Comment (eeedean) [inline pkgs/by-name/lm/lmstudio/darwin.nix:13]: Would you mind, putting the letter into a variable "revision"? As [pointed out at lmstudio](https://github.com/lmstudio-ai/lmstudio-bug-tracker/issues/25), that scheme seems to be meant to last.

File: pkgs/by-name/lm/lmstudio/package.nix [hunk 2]
```diff
@@ -5,13 +5,13 @@
...
   pname = "lmstudio";
-  version = "0.2.22";
+  version = "0.2.23";
   meta = {
     description = "LM Studio is an easy to use desktop app for experimenting with local and open-source Large Language Models (LLMs)";
     homepage = "https://lmstudio.ai/";
     license = lib.licenses.unfree;
     mainProgram = "lmstudio";
-    maintainers = with lib.maintainers; [ drupol eeedean ];
+    maintainers = with lib.maintainers; [ eeedean ];
```
Comment (eeedean) [inline pkgs/by-name/lm/lmstudio/package.nix:14]: If you feel like it, you could also add yourself as a maintainer.

**Review by eeedean — CHANGES_REQUESTED** [review]
Please adjust the commits by squashing them, where needed. Also take a look at the conventions for commit messages in [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#pull-request-template) as well as [README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#commit-conventions) of pkgs. 


The Commits should eventually look about like that: 

 * `maintainers: add cig0` or »your handle«
 * `lmstudio: remove drupol from maintainers`
 * `lmstudio: add cig0 to maintainers`
 * `lmstudio: 0.2.22 -> 0.2.23`

If you wouldn't want to be listed as maintainer, you can and should of course leave those commits out. 

Thanks for your contribution! 

Edit: I have no strong opinion about the `-Ubuntu-20.04`, too. Leave it, as you like 😺.

**Comment by cig0** [comment]
@eeedean, I addressed the issues you pointed out (thanks!); let me know what you think.

**Comment by eeedean** [comment]
It looks like you somehow pulled in commits, that don't belong to your branch. 
image


Also the changes are incomplete and are not correctly distributed over the commits: 

[maintainers: add cig0](https://github.com/NixOS/nixpkgs/pull/310914/commits/f8c7960cf2fadb08749ec90f4bf30baaa188f7c1) should only add your handle to [maintainers/maintainer-list.nix](https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix). Be aware, that this list must be ordered alphabetically, or one of the checks will fail. 
[lmstudio: remove drupol from maintainers](https://github.com/NixOS/nixpkgs/pull/310914/commits/da46df28dc984931a7765fc5457694430cc96d40) also does the actual update, which should go into [lmstudio: 0.2.22 -> 0.2.23](https://github.com/NixOS/nixpkgs/pull/310914/commits/1b09439e981304459597a3783a86ba81279fe92a). 

Best case would be to also add your own maintainer handle in a separate commit named like "lmstudio: adding cig0 to maintainers".

Also have a look at the failing check. 

<details> 
  <summary>If you want to figure it out yourself first, you can have a look at the failing test. Otherwise, I got the reason written in the spoiler here 😸 </summary>
   You call <pre>./darwin</pre> without the revision variable  <pre>then callPackage ./darwin.nix { inherit pname version meta; }</pre> 
but you should: <pre>then callPackage ./darwin.nix { inherit pname version »revision« meta; }</pre>
</details>

**Comment by cig0** [comment]
@eeedean I've been busy; I'm looking to push the changes today

**Comment by eeedean** [comment]
> @eeedean I've been busy; I'm looking to push the changes today

Did not mean to be pushy. Just love to close stuff and saw you commiting. 

Anyway: Thanks for your cooperation. Looking forward!

**Comment by cig0** [comment]
> > @eeedean I've been busy; I'm looking to push the changes today
> 
> Did not mean to be pushy. Just love to close stuff and saw you commiting.
> 
> Anyway: Thanks for your cooperation. Looking forward!

Oh, no, no, far from that! I was letting you know!

While cleaning up this mess, I noticed that the folks at LM Studio changed again the format of the download link for the Mac version, removing the revision identifier:

From https://github.com/NixOS/nixpkgs/pull/310914/files#diff-5932e72431b8e295a889e7c4defdd8aa02221a7f1103665b696e3b344dea3fb9L13 to `https://releases.lmstudio.ai/mac/arm64/0.2.24/latest/LM-Studio-0.2.24-arm64.dmg`.

We no longer need to add the new variable (until they change the URL again, I guess!).

In any case, I'll prepare the new PR and wait for your comments.

Thanks!

**Comment by cig0** [comment]
@eeedean, everyone: new PR here https://github.com/NixOS/nixpkgs/pull/314992

**Pull request closed without merging**